File: RB110/study/reverse_words.rb
=begin
def reverse_words(sentence)
  sentence.split.each {|word| word.reverse! }.join(' ')
end

sample_sentence = 'Hello World'
p reverse_words(sample_sentence) # => 'olleH dlroW'

Without using the reverse method

Algorithm:
  -Set a Result variable to an empty array
  -Split input string into an array
  -iterate over split input string
    -For each word in string
      -create a reversed word array
      -creat and index and set -1
      -starting at index -1 append to reverse word
      -decremet index by 1
    -Append rev word to result array
  -join array and return

=end


def reverse_words(sentence)
  result = []

  sentence.split.each do |word|
    index = -1
    rev_word = []
    loop do 
      rev_word << word[index]
      index -= 1
      break if rev_word.size >= word.size
    end
    result << rev_word.join
  end
  result.join(' ')
  
end

sample_sentence = 'Hello World'
p reverse_words(sample_sentence) # => 'olleH dlroW'

File: LS250/practice_problems/two_pointers/reverse_words.js
/*
You are given a sentence represented by a string str. Your objective is to reverse all the characters in 
each word of the sentence while ensuring that the case of each character remains unchanged. The spaces 
between words should be preserved as they are, and the overall order of the words in the sentence must 
not be altered.

You should solve the problem without using the Array.prototype.reverse method.

Input:
  - A string of letters
Output:
  - the string reversed

Problem:
  - Given a string we should output the string in reverse

Explicit:
  - Ensure the chars keep there case
  - ensure that the space inbetween the words are intact
  - The overall words in the sentences must not be altered

Implicit:
  - 

Data structure:
  - Start/end pointers
  - Split the word into an array

Algorithm:

  - Split the word into an array
  - Iterate over the array

  - Declare a start variable set to 0
  - Declare an end variable set to word length
  - While start is less then end
    - swap the char at index start with the char at index end
  - Return the revered word

  - Join the new array with ' ' 


*/

function charSwap(word) {
  let arr = word.split('');
  let start = 0;
  let end = word.length - 1;

  while (start < end) {
    [arr[start], arr[end]] = [arr[end], arr[start]]
    start++;
    end--;
  }

  return arr.join('');
}

function reverseWords(str) {
  let wordArr = str.split(' ');
  let reversed = [];

  wordArr.forEach(word => {
    reversed.push(charSwap(word));
  });

  return reversed.join(' ');
}


console.log(reverseWords("Hello World") === "olleH dlroW");
console.log(reverseWords("JavaScript is fun") === "tpircSavaJ si nuf");
console.log(reverseWords("Coding in the sun") === "gnidoC ni eht nus");
console.log(reverseWords("Launch School") === "hcnuaL loohcS");